Feature: adds bioprotocol proposal

lsh-0 opened the pull request.

lsh-0 committed:
adds bioprotocol proposal

giorgiosironi commented on src/model/bioprotocol.v1.yaml:
```diff
@@ -0,0 +1,25 @@
+        protocol_status:
+            description: "numeric code corresponding to the current status of a protocol"
+            type: "integer"
+            minimum: 0
```
anything better than a number? Is there a translation table from `0`, `1` to keywords?

lsh-0 replied:
there isn't

lsh-0 replied:
I've asked Dong to elaborate

lsh-0 replied:
Dong has gotten back to us. They actually switched from a boolean *to* an integer but his reply doesn't read like they intend to expand the number of different states any time soon.

I can store the integer and return a boolean value instead. When/if BP decide to add more statuses we can tweak things then.

giorgiosironi commented on src/model/bioprotocol.v1.yaml:
```diff
@@ -0,0 +1,25 @@
+            description: "section title, looks like 'Isolation of LDs and treatments'"
+            type: string
+        is_protocol:
+            type: boolean
```
always `false` in the example, what does it mean? An enumeration can be more extensible than a boolean e.g. `type`: `protocol`|`something-else` allows addition of other types.

lsh-0 replied:
this is a flag they use when they determine that a protocol that has been given to them isn't actually a protocol - eLife mixes materials and methods together in the same section with no means of differentiation, and there is some talk at the moment on how to best deal with that.

```
$ curl https://dev.bio-protocol.org/api/elife00005 | jq '.Protocols[].IsProtocol'`
true
true
true
true
true
true
```

I'll clarify with them, but I think what I'll do is exclude rows from being returned to the journal where `IsProtocol=false` so that a link isn't generated. And I'll exclude this field as well as it will then be redundant.

lsh-0 replied:
I've pinged Dong about it. It seems the most obvious and reasonable thing to do and they seem quite competent so I'll go ahead and make the changes. I can always revert if they start going crazy.

lsh-0 replied:
yes, Dong confirmed this is what is happening

giorgiosironi commented on src/model/bioprotocol.v1.yaml:
```diff
@@ -0,0 +1,25 @@
+        protocol_sequencing_number:
+            description: "section ID, looks like 's4-1'"
+            type: string
+        protocol_title:
```
is the `protocol_` prefix in these two fields necessary? Thinking about e.g. [podcast episodes](https://github.com/elifesciences/api-raml/blob/develop/src/snippets/podcast-episode.v1.yaml#L5) where we use specific terms like `number` without a prefix.

lsh-0 replied:
I can remove those

lsh-0 committed:
bioprotocol, updates definition and examples

lsh-0 commented:
raml changes are in and so are the code changes.

I'd like to merge this Friday if there are no further concerns

giorgiosironi commented on src/api.raml:
```diff
@@ -715,6 +715,27 @@ traits:
                                     displayName: Complete
                                     value: !include samples/collection/v2/complete.json
 
+/bioprotocol:
```
mark with [`@experimental`](https://github.com/elifesciences/api-raml/blob/40b376afaa33083a9a86dd3516f171e091e7ffd0/adr/0002-use-raml-annotation-experimental.md) so that we can make amendments later without versioning

lsh-0 replied:
done

thewilkybarkid commented on src/api.raml:
```diff
@@ -715,6 +715,27 @@ traits:
+                        application/vnd.elife.bioprotocol+json;version=1:
+                            schema: !include ../dist/model/bioprotocol.v1.json
+                            examples:
+                                complete:
```
I'd add an `empty` example too.

lsh-0 replied:
done

thewilkybarkid commented on src/model/bioprotocol.v1.yaml:
```diff
@@ -0,0 +1,21 @@
+title: BioProtocol
+type: array
+items:
+    type: object
```
Everything should be `required`?

lsh-0 replied:
everything is now required

thewilkybarkid commented on src/model/bioprotocol.v1.yaml:
```diff
@@ -0,0 +1,21 @@
+items:
+    type: object
+    properties:
+        sequencing_number:
```
Using camelCase elsewhere. Since this is a section ID, call it `sectionId`?

lsh-0 replied:
done, is now camelCase

lsh-0 replied:
renamed sequencingNumber to sectionId

thewilkybarkid commented on src/model/bioprotocol.v1.yaml:
```diff
@@ -0,0 +1,21 @@
+        sequencing_number:
+            description: "section ID, looks like 's4-1'"
+            type: string
+        title:
```
Not needed for machines, but probably useful for humans.

thewilkybarkid commented on src/model/bioprotocol.v1.yaml:
```diff
@@ -0,0 +1,21 @@
+            type: string
+        status:
+            description: "numeric code corresponding to the current status of a protocol"
+            type: "boolean"
```
Numeric or boolean? If it a boolean, name needs to match (`status` doesn't work), otherwise an enum should be used.

lsh-0 replied:
it was changed to boolean from an integer. I'll change the description.

Why doesn't 'status' work? is it a reserved keyword or something?

thewilkybarkid replied:
Doesn't work as in I don't know what `status: false` means (same with `status: 1`).

lsh-0 replied:
how about `has_protocol` ?

lsh-0 replied:
this boolean value indicates whether a 'request protocol' link should be created or a 'view protocol' link should be created. `status` seems fine to me, `has_protocol` doesn't seem so bad either

lsh-0 replied:
I'm keeping `status` and have improved the description. It's clearer now.

lsh-0 committed:
bioprotocol, struct more closely follows other structs now
adds 'empty' example
adds 'experimental' nesting
changed struct naming to camelCase

lsh-0 committed:
renamed sequencingNumber to sectionId

lsh-0 commented:
review feedback has been incorporated. the output structure more closely resembles the lists of results of everything else in the api now.
